### ISSUE TITLE
bigint.divexact(const ref n: bigint, d: integral)  output unexpected for a particular case.

### ISSUE THREAD

**Issue by avneet14027** [chapel-lang/chapel#10345]
When the divexact(const ref n: bigint, d: integral) function is called on a bigint variable, and the arguments are a bigint array and an integer, then the output is the quotient for the last element of the array divided by the integer given as argument. 

For eg: 
```var array: [1..5] bigint();
var d = new bigint(5);
var x = new bigint();
array[1] = 79;
array[2] = 30;
array[3] = 33;
array[4] = 60;
array[5] = 50;

x.divexact(array,d);
writeln(array);
writeln(x);
writeln(d);
```
For the above code, the output is the following:
```
79 30 33 60 50 // array
10 // last element divided by divisor (50/5)
5 // divisor
```
Here, x is the variable on which the function is called, and d is the divisor, array is a bigint array. The output which is 5, stored is x, which is 50/5, i.e. last element of the array divided by the integral divisor.

This functionality isn’t obvious, unlike the cases,
- when two bigint arrays given as arguments, are divided with each other and the result is stored in the bigint array on which the function is called.
Or 
- Two bigint variables given as arguments, are divided with each other, and the result is stored in the bigint variable on which the function is called.

Since the aforementioned functionality isn’t obvious, a compiler warning should be generated for the same.

**Comment by LouisJenkinsCS** [comment]
Running the code on [TIO](https://tio.run/##Vc1BC8IgGMbxu5/iPS4IaZlFjS7d@gxjB5svJgwJs80@vanLWDd//0e0v4sHDiG8nggXra7GoULbEDIKC8Ja8YYTtDWlvIObVtq4Jk8SzmBw@raKr@bs/3Os@Y227uJwOBZuE9mmkGWywl3i/rfyRB5JPJV6RC96V@VpLeMHk9UOBzOXhf3inO6F8AE) shows something like undefined behavior? Output for release 1.17 seems to be 

```
79 30 33 60 50
7378697629483820653
5
```

**Comment by LouisJenkinsCS** [comment]
Looking at documentation for GMP's [`exact` and `exact_ui`](https://gmplib.org/manual/Integer-Division.html), which is presumably the equivalent of `bigint.exactdiv` method: "Set q to n/d. These functions produce correct results only when it is known in advance that d divides n."

Also, I believe that this behavior is expected. Due to [implicit promotion](https://chapel-lang.org/docs/master/users-guide/datapar/promotion.html#examples-users-guide-datapar-promotion-chpl), the array of `bigint` turns into a sequence of function calls with each element of the array being an argument. [Example](https://tio.run/##S85ILEjN@f@/oCg/WaGgKDOvRKPCSgFIaSpUc3GWF2WWpObkaSi555coJBall@am5pVYKSjpKFRoWnPVcnGVJRYBxYsSKxWsFKIN9fRMY0F6rbnAYtGGsQq2CuaWMK4RiGtsAOMag7nGMK4JiGsGlzUFcU2BXC6Iq8Cimtb//wMA) of implicit promotion of array as argument.

**Comment by bradcray** [comment]
Hi @avneet14027 : Thanks for filing this issue.  I think Louis is correct about promotion being responsible for what's going on here.

[For those reading this issue who don't want to read the [users guide link](https://chapel-lang.org/docs/users-guide/datapar/promotion.html) that Louis provided, In Chapel, in general, any function or method call expecting an argument of type `t` that's passed an array of `t` instead results in that function/method being called for all elements of the array in parallel].

Specifically, in this case, the call:

```chapel
x.divexact(array,d);
```

is essentially equivalent to:

```chapel
forall a in array do
  x.divexact(a,d);
```

This explains the instability that Louis mentions (and that I'm observing as well), since multiple parallel calls are all trying to write to a single variable `x`, causing a race condition.

In contrast, the case you mention in which the method is called on an array of values:

```chapel
array2.divexact(array,d);
```

is equivalent to:

```chapel
forall (a, a2) in zip(array, array2) do
  a2.divexact(a,d);
```

Since this kind of behavior could happen in any promoted case, I don't think we'd want to put a special error in for this specific function signature; but it raises an interesting question about whether promoted function calls in which non-promoted actual arguments whose corresponding formals' intents suggest that they'll be modified (i.e., `ref`, `inout`, `out`) should result in a warning if they're not promoted (passed an array value) in an otherwise promoted call.

Connecting that back to your example, divexact has the signature:

```chapel
  proc bigint.divexact(const ref n: bigint, const ref d: bigint) { ... }
```

which can be thought of as being similar to this procedure, if we make the 'this' argument explicit:

```chapel
  proc divexact(ref this: bigint, const ref n: bigint, const ref d: bigint) { ... }
```

So if we think of your function call as being:

```chapel
divexact(<non-array>, <array>, <non-array>);
```

then we see that the `ref` intent on the first argument suggests that the `this` actual argument is going to be modified, so it could be an indication of a problem if an array argument isn't passed in to match the promotion caused by the actual passed to `n`.

I think such a warning has merit, though it gets a bit into questions about how good a job the Chapel compiler currently does in detecting whether `this` arguments specifically are modified / modifiable or not (something we were talking about in a meeting just last week as it turns out).  I'm tagging @benharsh and @mppf on this issue because the three of us spent a lot of time talking about `this` intents a year or so back.

If others think that such a warning has merit, I think we should fire up a new issue proposing it in place of this one.

**Comment by mppf** [comment]
I'm for a warning in this case & in fact I asked @avneet14027 to create this issue to start that discussion. We could certainly make another issue if that seems important.

**Comment by bradcray** [comment]
I think if there's agreement in generating the warning, then a more specific issue whose title refers to something like "generate warnings for [potential] modification of non-promoted arguments in promoted calls" would be easier to find than the current one (we could completely change the title and description of this issue, but then the rest of these comments will seem out of place...).

**Comment by mppf** [comment]
@bradcray - see #10412

**Comment by bradcray** [comment]
@avneet14027 : With issue #10412 now filed as a more general request that would have solved your issue, would you be OK closing this issue, or do you think it should stay open for other reasons?

**Comment by avneet14027** [comment]
@bradcray Yes, that solves it for me, I'll be closing it now. Thank you!

**Comment by bradcray** [comment]
Thanks for reporting it!
